lib/ecto/adapters/pool.ex

```diff
@@ -152,198 +145,217 @@ defmodule Ecto.Adapters.Pool do
     catch
-      class, reason ->
+      # If any error leaked, it should be a bug in Ecto.
+      kind, reason ->
         stack = System.stacktrace()
         break(ref, timeout)
-        :erlang.raise(class, reason, stack)
+        :erlang.raise(kind, reason, stack)
+    else
+      res ->
+        close_transaction(ref, Process.get(ref), timeout)
+        res
+    after
+      Process.delete(ref)
     end
   end
 
-  defp do_run(pool_mod, pool, timeout, fun) do
-    case checkout(pool_mod, pool, timeout) do
-      {:ok, worker, conn, time} ->
-        try do
-          fun.(conn, time)
-        catch
-          class, reason ->
-            stack = System.stacktrace()
-            pool_mod.break(pool, worker, timeout)
-            :erlang.raise(class, reason, stack)
-        else
-          res ->
-            pool_mod.checkin(pool, worker, timeout)
-            {:ok, res}
-        end
-      {:error, _} = error ->
-        error
+  defp inner_transaction(ref, conn, fun) do
+    try do
+      fun.(:already_open, ref, conn, nil)
+    catch
+      kind, reason ->
+        stack = System.stacktrace()
+        taint(ref)
+        :erlang.raise(kind, reason, stack)
     end
   end
 
-  defp checkout(pool_mod, pool, timeout) do
-    case pool_mod.checkout(pool, timeout) do
-      {:ok, _worker, _conn, _time} = ok ->
-        ok
-      {:error, reason} = error when reason in [:noproc, :noconnect] ->
-        error
-      {:error, err} ->
-        raise err
-    end
+  defp close_transaction({__MODULE__, pool_mod, pool}, %{conn: _, worker: worker}, timeout) do
+    pool_mod.close_transaction(pool, worker, timeout)
+    :ok
   end
 
-  defp transaction(pool_mod, pool, ref, timeout, fun) do
-    case open_transaction(pool_mod, pool, timeout) do
-      {:ok, info, time} ->
-        try do
-          do_transaction(ref, info, time, fun)
-        after
-          info = Process.delete(ref)
-          close_transaction(pool_mod, pool, info, timeout)
-        end
-      {:error, _} = error ->
-        error
-    end
+  defp close_transaction(_, %{}, _) do
+    :ok
   end
 
-  defp do_transaction(ref, %{depth: depth, conn: conn} = info, time, fun) do
-    depth = depth + 1
-    _ = Process.put(ref, %{info | depth: depth})
+  @doc """
+  Executes the given function giving it the ability to rollback.
+
+  Retunrs `{:ok, value}` if no transaction ocurred,
+  `{:error, value}` if the user rolled back or
+  `{:raise, kind, error, stack}` in case there was a failure.
+  """
+  @spec with_rollback(ref, (() -> return)) ::
+        {:ok, return} | {:error, term} | {:raise, atom, term, Exception.stacktrace}
+        when return: var
+  def with_rollback(ref, fun) do
     try do
-      {:ok, fun.(ref, conn, depth, time)}
-    after
-      case Process.put(ref, info) do
-        %{conn: _} ->
-          :ok
-        %{} ->
-          _ = Process.put(ref, Map.delete(info, :conn))
-          :ok
+      value = fun.()
+      if untaint(ref) do
```

Comment:
Actually, we could call it on `after`. It would be better. I will clean it up tomorrow.